keps/74-support-argo-workflow/README.md

```diff
@@ -0,0 +1,350 @@
+ ✔ loops-644ch                              loop-example-depth-1
+ └─┬─✔ loop-example-depth-2-2               loop-example-depth-2
+   │ └─┬─✔ print-message-loop(0:depth-2-1)  print-message         loops-644ch-print-message-1796012204  4s
+   │   └─✔ print-message-loop(1:depth-2-2)  print-message         loops-644ch-print-message-1116167650  6s
+   ├─✔ print-message(0:depth-1-1)           print-message         loops-644ch-print-message-413467513   5s
+   └─✔ print-message(1:depth-1-2)           print-message         loops-644ch-print-message-3356863351  5s
+```
+
+##### Example 3 (Workflow with Single Container Template)
+
+We create a workload for the single container template. For example:
+```
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: hello-
+spec:
+  entrypoint: main
+  templates:
+    - name: main
+      plugin:
+        hello: { }
+
+# argo get hello-jtlcw
+...
+STEP            TEMPLATE  PODNAME  DURATION  MESSAGE
+ ◷ hello-jtlcw  main
+```
+
+#### How to suspend a workflow step by step
+
+It is hard for users to add suspend template manually in a workflow before each
+leaf template. Some tools are needed to help them to do this.
+We introduce three ways to manage the workflow. Responsebilities are different for the
+workflow-controller and kueue-controller in two ways.
+
+1. Give users a CLI to help users modifying workflows to add a specific suspend template for each step.
+When the workflows are suspended on this special suspend template, the job-controller in Kueue
+create workloads for the next step. Modification of workflow-controller is not needed for
+this way, so that it is easy to iterate, and no need to manage the version of argo and kueue.
+By in this way, users can modify their workflows to skip waiting in kueue, which maybe is not 
+acceptable for some users.
+
+2. Integrated Suspend Capability: We propose introducing a new specification field within workflows, such 
+as suspendBySteps. If workflow.spec.suspendBySteps is set to true, the workflow-controller automatically 
+injects a special suspend template into each stepGroup. The kueue's job-controller monitors these and 
+generates workloads for the following steps. Upon workload admission, the suspension step is marked as 
+completed.
+
+3. (Recommended) Kueue Webhook Enhancement: A new webhook is added within Kueue to intercept pod creations in the cluster. 
+This webhook verifies if the incoming pods are governed by a workflow and if the workflow carries the label 
+`kueue.x-k8s.io/queue-name`. When these conditions are met, scheduling gates are appended to the pods. The 
+job-controller in Kueue subsequently organizes these pods into groups (identifiable within the workflow's 
+status) and creates corresponding workloads for each group. Following workload acceptance, the scheduling 
+gates are removed from the pods, enabling their scheduling and execution.
+
```

Comment:
How can we push all steps and leaves sequentially in this approach?
IIUC, when others push other Jobs into same queue during Workflow preprocessing executed, the following Workflow step needs to wait until the other irrelevant Job is terminated.

Let's imagine the situation where we push a ArgoWorkflow and a Job into the same queue:

1. Push ArgoWorkflow into queue
2. Creation of Workload for the first Argoworkflow step
3. Admit the first Workflow step Pod
4. Push an irrelevant Job into queue
5. Finish the first Workflow step Pod
6. Admit an irrelevant Job.
7. The second Workflow step Pod waits until the irrelevant Job finish.

Reply:
That is what I call Admission Reuse. Maybe we should support relevant Job to reuse the resources to avoid waiting for irrelevant. Maybe this should be a standard API in Kueue, not just for workflow.

Reply:
This KEP will not discuss the solution for this problem. And maybe the solution can be consider with this issue. 
#1243

Reply:
No, I did not mean the feature.
Couldn't we create all Workloads with active=false once the workflow is created?
So, the creation of all Workloads for all steps allows us to enqueue all steps sequentially, and we can avoid the reservation of the whole of WorkflowTemplate.

Reply:
Conditional execution is supported in most workflow managers, so we can not determine whether the subsequent steps will run or how many replicas will be created until the step node is created.